x-pack/legacy/plugins/integrations_manager/common/types.ts

```diff
@@ -33,3 +45,19 @@ export interface IntegrationInfo {
     };
   };
 }
+
+// the public HTTP response types
+export type IntegrationList = IntegrationListItem[];
+
+export type IntegrationListItem = Installed<RegistryListItem> | NotInstalled<RegistryListItem>;
```

Comment:
The idea here is the same as in:

https://github.com/elastic/kibana/blob/d116949c6163933261e43d4f695317a9ed5d88bd/x-pack/legacy/plugins/integrations_manager/server/integrations.ts#L204-L214

LMK if there’s a better/idiomatic/whatever way of doing this.

Reply:
I actually would not have written this like you did... but I like how you did it and might adopt it in my own code 😊

Reply:
I think this looks pretty good already. For improved inference and readability it might be beneficial to store the installed type as a property, as in

```typescript
interface Installed<T> {
  status: 'installed';
  value: T;
  savedObject: InstallationSavedObject;
}

interface NotInstalled<T> {
  status: 'not_installed';
  value: T;
}

type MaybeInstalled<T> = Installed<T> | NotInstalled<T>;

export type IntegrationInfo = MaybeInstalled<RegistryPackage>;
```

That way the set of properties of the value is preserved and cleanly separated from the installation metadata (e.g. for iteration or serialization). As a pattern It also has the advantage that it's agnostic to the value's type.

Reply:
@weltenwort Thanks for the feedback. I considered something just like that. I think it’s probably the better/cleaner pattern. 

I’ll keep thinking but it seems like a change worth making.